File: ios/Anemone SDK/model/tenant/Tenant.swift
//
//  Tenant.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 9/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//
// swiftlint:disable identifier_name line_length colon

import Foundation

public typealias TenantId = Id
 
public struct Tenant {
    public let id: TenantId
    public let name: String
    public let description: String?
    public let address: Address
    public let images: [String]
    public let resources: [TenantResource]
    public let sportIds: [SportId]
    public let type: TenantType?
    public let bookingType: BookingType
    public let cancelationPolicies: [CancelationPolicy]?
    public let openingHours: [String: TimeRange]?
}

public extension Tenant {
    func fulfillsFilters(_ filters: [PropertyId: [PropertyOptionId]]?) -> Bool {
        guard let filters = filters else { return true }
        for resource in resources {
            if resource.fulfillsFilters(filters) {
                return true
            }
        }
        return false
    }

    func resources(sportId: SportId) -> [TenantResource] {
        resources.filter { $0.sportId == sportId }
    }
}

extension Tenant: JSONMappable {
    public init(json: JSONObject) throws {
        id = try TenantId(json.getAny("tenant_id"))
        name = try json.optString("tenant_name") ?? json.getString("name")
        address = try Address(json: json.getJSONObject("address"))
        images = try json.getStringArray("images")
        description = try? json.getString("description")
        sportIds = try json.getAnyArray("sport_ids").map { try SportId($0) }
        type = (try? json.getString("tenant_type")).let { TenantType.fromRawValue($0) }
        resources = json.optJSONArray("resources")?.flatMap { obj in try? TenantResource(json: obj) } ?? []
        let bookingTypeStr = try json.getString("booking_type")
        guard let bookingType = BookingType.fromRawValue(bookingTypeStr) else {
            throw AnemoneException.jsonInvalidFormat(key: "booking_type")
        }
        self.bookingType = bookingType
        cancelationPolicies = try? json.getJSONArray("cancelation_policies").flatMap { try CancelationPolicy(json: $0) }
        if let openingJson = json.optJSONObject("opening_hours") {
            var hours = [String: TimeRange]()
            openingJson.keys().forEach { day in
                if
                    let from = try? Time(openingJson.getJSONObject(day).getString("opening_time")),
                    let to = try? Time(openingJson.getJSONObject(day).getString("closing_time"))
                {
                    hours[day] = TimeRange(from: from, to: to)
                }
            }
            self.openingHours = hours
        } else {
            self.openingHours = nil
        }
    }
}

extension Tenant: JSONSerializable {
    public func toJson() -> JSONObject {
        let json = JSONObject()
        json.setString("tenant_id", id.value)
        json.setString("tenant_name", name)
        json.setObject("address", address.toJson())
        json.setArray("images", images)
        json.setString("description", description)
        json.setArray("sport_ids", sportIds.map { $0.value })
        json.setString("booking_type", bookingType.description)
        return json
    }
}

extension Tenant: Equatable { }
public func == (lhs: Tenant, rhs: Tenant) -> Bool {
    lhs.id == rhs.id
}

extension Tenant: Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/model/tenant/Tenant.kt
package com.anemonesdk.model.tenant

import com.anemonesdk.general.json.JSONMappable
import com.anemonesdk.general.json.JSONObject
import com.anemonesdk.model.config.PropertyId
import com.anemonesdk.model.config.PropertyOptionId
import com.anemonesdk.model.config.SportId
import com.anemonesdk.model.generic.Address
import com.anemonesdk.model.generic.Id
import com.playtomic.foundation.extension.tryOrNull
import com.playtomic.foundation.model.Time
import com.playtomic.foundation.model.TimeRange
import org.json.JSONException

/**
 * Created by mgonzalez on 20/12/16.
 */

open class TenantId(value: Any) : Id(value = value)

class Tenant : JSONMappable {
    val id: TenantId
    val name: String
    val description: String?
    val address: Address
    val images: List<String>
    val resources: List<TenantResource>
    val sportIds: List<SportId>
    val type: TenantType?
    val bookingType: BookingType
    val cancelationPolicies: List<CancelationPolicy>?
    val properties: Map<String, String>?
    val openingHours: Map<String, TimeRange>?

    constructor(json: JSONObject) : super(json) {
        id = TenantId(json.getAny("tenant_id"))
        name = json.optString("tenant_name") ?: json.getString("name")
        address = Address(json.getJSONObject("address"))
        images = json.getStringArray("images")
        description = json.optString("description")
        sportIds = json.getAnyArray("sport_ids").map { SportId(it) }
        type = json.optString("tenant_type")?.let { TenantType.fromRawValue(it) }
        resources = json.optJSONArray("resources")?.flatMap { obj -> tryOrNull { TenantResource(obj) } } ?: listOf()
        val bookingTypeStr = json.getString("booking_type")
        this.bookingType = BookingType.fromRawValue(bookingTypeStr)
            ?: throw JSONException("Invalid booking type $bookingTypeStr")
        cancelationPolicies = tryOrNull { json.getJSONArray("cancelation_policies").flatMap { CancelationPolicy(json = it) } }
        properties = (tryOrNull { json.getJSONObject("properties") })?.asMap()
        val openingJson = json.optJSONObject("opening_hours")
        if (openingJson != null) {
            val hours = mutableMapOf<String, TimeRange>()
            openingJson.keys().forEach { day ->
                val from = tryOrNull { Time(openingJson.getJSONObject(day).getString("opening_time")) }
                val to = tryOrNull { Time(openingJson.getJSONObject(day).getString("closing_time")) }
                if (from != null && to != null) {
                    hours[day] = TimeRange(from = from, to = to)
                }
            }
            this.openingHours = hours
        } else {
            this.openingHours = null
        }
    }

    constructor(
        id: TenantId,
        name: String,
        description: String,
        address: Address,
        images: List<String>,
        resources: List<TenantResource>,
        sportIds: List<SportId>,
        type: TenantType,
        bookingType: BookingType,
        cancelationPolicies: List<CancelationPolicy>?,
        properties: Map<String, String>?,
        openingHours: Map<String, TimeRange>?
    ) {
        this.id = id
        this.name = name
        this.description = description
        this.address = address
        this.images = images
        this.resources = resources
        this.sportIds = sportIds
        this.type = type
        this.bookingType = bookingType
        this.cancelationPolicies = cancelationPolicies
        this.properties = properties
        this.openingHours = openingHours
    }

    fun fulfillsFilters(filters: Map<PropertyId, List<PropertyOptionId>>?): Boolean {
        if (filters == null) {
            return true
        }

        for (resource in resources) {
            if (resource.fulfillsFilters(filters)) {
                return true
            }
        }
        return false
    }

    fun resources(sportId: SportId) = resources.filter { it.sportId == sportId }

    override fun equals(other: Any?) =
        (other is Tenant) && other.id == id

    override fun hashCode(): Int = id.hashCode()

}
